### DISCUSSION TITLE
Homebrew libraries for deployment on older macOS versions

### DISCUSSION THREAD

**Discussion by cme** [Homebrew/discussions#370] [Everyday usage]
Hi!

Over at the [Hydrogen](http://github.com/hydrogen-music) project, we're encountering some issues building releases with Homebrew dependencies.

The build uses MACOSX_DEPLOYMENT_TARGET to build the release, and uses Homebrew to gather dependencies (including Qt) for packaging. However, homebrew seems to *always* install bottles for the current host operating system, and I don't see any options to encourage it to install older OS bottles so that the libraries may be used in packaging for older macOS versions.

It seems we can manually force this by explicitly downloading older OS bottles and installing them, but this seems error-prone; or we can rebuild from source with MACOSX_DEPLOYMENT_TARGET set, but this would create prohibitively long-running builds, and seems wasteful given the bottles are likely available.

Any advice?

Many thanks! :)

**Comment by jonchang** [comment]
> The build uses MACOSX_DEPLOYMENT_TARGET to build the release, and uses Homebrew to gather dependencies (including Qt) for packaging. However, homebrew seems to always install bottles for the current host operating system, and I don't see any options to encourage it to install older OS bottles so that the libraries may be used in packaging for older macOS versions.

This is intentional. There can be subtle incompatibilities when installing bottles built for older verisons of macOS. It usually works but there are many cases where it doesn't. See https://github.com/Homebrew/brew/issues/10127.

> It seems we can manually force this by explicitly downloading older OS bottles and installing them, but this seems error-prone; or we can rebuild from source with MACOSX_DEPLOYMENT_TARGET set, but this would create prohibitively long-running builds, and seems wasteful given the bottles are likely available.

This is not something we support unfortunately, (using Homebrew-built binaries for redistribution in an .app). I agree that building Qt is a pain, but your best bet is to probably to use a VM of an older version of macOS to build your software.

**Comment by btsimonh** [comment]
I have a very similar situation.  I have to create distributable binaries for MACOS (packaged), but the only MAC I have is quite an early Mini.  I shoe-horned Mojave on there (patched), mainly to test signing, etc. and all has been fine....
Until...  Previously I was building everything from source.   Took a shortcut to using brew to install deps before building Opencv - and hey presto, Illegal instruction - because I'm getting Mojave bottles.  The 'thing' I need is available for High Sierra, which might work, but no way to lie to brew about the OS?
